Add some more tests

edorfaus opened the pull request:
This adds tests for more aspects of the hash, mainly testing the happy path for various inputs, but also a few other things. It also refactors the existing tests a bit, to make things easier.

Most of the new hashes were produced using the ed2k_hash tool (which produces blue), so they _should_ be correct. The exception is the hashes for an empty input, for which ed2k_hash errors out; I don't have any outside verification that my assumptions for them are correct, as they were created manually using MD4 in the Go playground.

The last commit adds two tests that currently fail; both of them are for inputs that are smaller than one chunk (but not empty), so I expect that fixing one will also fix the other. The equivalent tests for inputs that are between one and two chunks, or above two, already pass. This PR does not attempt to fix this problem, it only exposes it.

edorfaus commented:
Oops, found a bug in the failing tests - they weren't actually using the hashes they should. I've updated the PR to fix that. However, those tests still fail, even with the correct hashes.

edorfaus commented:
After looking at the AniDB wiki page again, I'm re-thinking my assumptions regarding the red hash for empty input.

Currently, that test assumes that blue should be `MD4("")` while red should be `MD4(MD4(""))`, which is also what the implementation produces. Basically, that red should have the same problem with 0 as for any other multiple of the chunk size.

However, I now think that the first block of the pseudo code on the wiki says that it should be `MD4("")` for both versions, since it says to _return_ there, without moving on to the rest of the pseudo code, and thus never reaching the red block.

If that's correct, then the empty-input test should be fixed, and should actually be failing, not passing, with the current implementation.

zorchenhimer commented:
So I was looking at this the other day and some stuff didn't add up for me.  Can you detail how you found the specific hashes to check against?

edorfaus commented:
The one-chunk and two-chunk hashes are from your original code (so, from the AniDB page). For most of the rest, I created zero-filled files using `truncate -s <size> <filename>` and then ran `ed2k_hash` on the resulting files.

I got that ed2k_hash program from the `ed2k-hash` package in Debian, version `0.4.0+ds-2`, which says it's from http://ed2k-tools.sourceforge.net/ed2k_hash.shtml -- this link is also on the AniDB page, which also mentions that 0.4.0 of ed2k_hash produces the blue hash.

Since the blue and red hashes should only be different on multiples of the chunk size, and most of the tests I added are not at such multiples, I used those hashes for testing both blue and red.

This leaves only the hash for the empty input, for which the ed2k_hash tool gave an error instead of a hash; so at the time, I used the Go playground to create the hashes for myself. I don't have the original code anymore, but to recreate it, it went something like this:
```go
blue := md4.New().Sum(nil)
fmt.Printf("blue: %x\n", blue)
h := md4.New()
h.Write(blue)
fmt.Printf("red: %x\n", h.Sum(nil))
```
In other words, blue is the MD4 of an empty input, and then for red, it basically appends the MD4 of nothing to an empty hashlist, and hashes that. In other words, it does what the red part of the AniDB pseudocode says, which I think I did because I assumed that 0 counts as a multiple of the chunk size (which it technically is), and that's what made sense to me at the time. Apparently I was a bit confused, since the pseudocode doesn't really allow this.

---
Also, I have now grabbed two other ed2k hashing tools that were mentioned on the AniDB page (ed2k by Stephane D'Alu, and jacksum), both of which it says uses the red hashes, and ran them on the same files. They both give the same checksums, which match the known red sums, and match the blue ones (from ed2k_hash) for all the non-multiple-of-chunksize files.

They also don't error out on the empty file, and for that they show the MD4-of-nothing hash (what the playground code calls "blue"), which confirms my suspicion that my test for empty input is incorrect with regards to the red hash. So I should fix that test.

By the way, while this PR doesn't try to fix the bugs, I do have code that does (though I now need to fix it for empty inputs, since I wrote it for the different red hash). I didn't include it here both because it wasn't written yet, and because the fixes wasn't the main idea in that code - it involves a significant refactor that I'm not sure you'll want. :P Thus I figured it was better left to a separate PR. Also, since making it, I realized I hadn't asked if this project is one you actually _want_ code contributions for... so... yeah. Sorry if it isn't.

edorfaus commented:
OK, I've now fixed the tests, and rebased to keep the history cleaner - I moved the empty-input test to the last commit, which now holds the three failing tests for small inputs. (I pushed the fixes as a separate commit first, in case you want to see what actually changed in the code.)

zorchenhimer commented:
That works for me, thanks!

I'll probably see if i can fix the bad hashes on stream tomorrow, but writing tests is a PITA, so you're more than welcome to fix those, LOL.
